hashicorp-vault/main.tf

```diff
@@ -0,0 +1,224 @@
+  count = var.s3_storage ? 1 : 0
+  statement {
+    sid    = "VaultOwn"
+    effect = "Allow"
+
+    actions = [
+      "kms:*"
+    ]
+
+    resources = [aws_kms_key.this.0.arn]
+  }
+}
+
+# TODO test usage of iam role and try to read credentials from secret
+resource "aws_iam_user" "this" {
+  count = var.s3_storage ? 1 : 0
+  name  = "${var.cluster_name}-vault"
+  path  = "/vault/"
+
+  tags = var.tags
+}
+
+resource "aws_iam_access_key" "this" {
+  count = var.s3_storage ? 1 : 0
+  user  = aws_iam_user.this.0.name
+}
+
+resource "local_file" "this" {
+  count    = local.argocd_enabled
+  content  = yamlencode(local.app)
+  filename = "${var.argocd.path}/${local.name}.yaml"
+}
+
+
+resource "helm_release" "this" {
+  count      = 1 - local.argocd_enabled
+  depends_on = [kubernetes_namespace.this]
+  name       = local.name
+  repository = local.repository
+  chart      = local.chart
+  namespace  = local.namespace
+
+  dynamic "set" {
+    for_each = local.conf
+
+    content {
+      name  = set.key
+      value = set.value
+    }
+  }
+}
+
+locals {
+  name           = var.chart_name
+  repository     = "https://helm.releases.hashicorp.com"
+  chart          = "vault"
+  version        = var.chart_version
+  namespace      = var.chart_namespace
+  argocd_enabled = length(var.argocd) > 0 ? 1 : 0
+
+  app = {
+    "apiVersion" = "argoproj.io/v1alpha1"
+    "kind"       = "Application"
+    "metadata" = {
+      "name"      = local.name
+      "namespace" = var.argocd.namespace
+    }
+    "spec" = {
+      "destination" = {
+        "namespace" = local.namespace
+        "server"    = "https://kubernetes.default.svc"
+      }
+      "project" = "default"
+      "source" = {
+        "repoURL"        = local.repository
+        "targetRevision" = local.version
+        "chart"          = local.chart
+        "helm" = {
+          "parameters" = values({
+            for key, value in local.conf :
+            key => {
+              "name"  = key
+              "value" = tostring(value)
+            }
+          })
+        }
+      }
+      "syncPolicy" = {
+        "automated" = {
+          "prune"    = true
+          "selfHeal" = true
+        }
+      }
+    }
+  }
+
+  conf = merge(local.conf_defaults, var.conf)
+  conf_defaults = merge(
+    var.s3_storage ? {
+      "server.standalone.config" = <<EOT
+${local.vault_conf}
+storage "s3" {
+  access_key = "${aws_iam_access_key.this.0.id}"
+  secret_key = "${aws_iam_access_key.this.0.secret}"
+  bucket     = "${var.s3_bucket_name}"
+  region     = "${var.s3_bucket_region}"
+}
+EOT
+    } : {},
+    var.file_storage ? {
+      "server.volumes[0].name"                            = var.file_storage_name,
+      "server.volumes[0].persistentVolumeClaim.claimName" = var.file_storage_pvc_name,
+      "server.standalone.config"                          = <<EOT
+${local.vault_conf}
+storage "file" {
+  path = "/vault/data"
+}
+EOT
+    } : {},
+    {
+  })
+
+  vault_conf = <<EOT
+ui = true
```

Comment:
If we set UI = true, we need to add some properties for ingress URL and annotations like a cert-manager

Reply:
ok, I'll remove this. Seems like it's redundant.